File: python/runner.py
import os
import re
from typing import Any, Dict, List, Optional, Set

VERSION_REGEX = re.compile(r"^(\d+)(?:_(.+))?\.sql$")


class ResultDict(dict):
    """A dictionary that also allows attribute access (e.g. res.success, res.migrations)."""
    def __getattr__(self, key: str) -> Any:
        try:
            return self[key]
        except KeyError:
            raise AttributeError(f"'ResultDict' object has no attribute '{key}'")

    def __setattr__(self, key: str, value: Any) -> None:
        self[key] = value


def run_on_connection(conn: Any, migrations_dir: str) -> ResultDict:
    """
    Applies all pending SQL migrations found in migrations_dir against any standard
    DB-API 2.0 database connection (e.g. sqlite3, psycopg2, pymysql).
    
    Each migration is executed within a transaction, and the schema_migrations table
    is automatically created and updated.
    """
    try:
        cursor = conn.cursor()

        # 1. Initialize tracking table
        init_sql = """
        CREATE TABLE IF NOT EXISTS schema_migrations (
            version VARCHAR(255) PRIMARY KEY,
            applied_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
        );
        """
        cursor.execute(init_sql)
        conn.commit()

        # 2. Query applied migrations
        cursor.execute("SELECT version FROM schema_migrations;")
        applied_set: Set[str] = {str(row[0]) for row in cursor.fetchall()}

        # 3. Read migration files
        if not os.path.exists(migrations_dir):
            raise FileNotFoundError(f"Migrations directory not found: {migrations_dir}")

        entries = []
        for filename in os.listdir(migrations_dir):
            if not filename.endswith(".sql"):
                continue
            match = VERSION_REGEX.match(filename)
            if match:
                version = match.group(1)
                name = match.group(2) or filename[:-4]
                entries.append((version, name, filename))

        entries.sort(key=lambda x: x[0])

        applied: List[ResultDict] = []
        for version, name, filename in entries:
            if version in applied_set:
                continue

            file_path = os.path.join(migrations_dir, filename)
            with open(file_path, "r", encoding="utf-8") as f:
                sql_content = f.read()

            try:
                if sql_content.strip():
                    cursor.execute(sql_content)

                record_sql = f"INSERT INTO schema_migrations (version) VALUES ('{version}');"
                cursor.execute(record_sql)
                conn.commit()

                applied.append(ResultDict({"version": version, "name": name}))
            except Exception as e:
                conn.rollback()
                raise e

        return ResultDict({
            "success": True,
            "applied": len(applied),
            "migrations": applied,
            "error": None,
        })
    except Exception as e:
        return ResultDict({
            "success": False,
            "applied": 0,
            "migrations": [],
            "error": str(e),
        })


def status_on_connection(conn: Any, migrations_dir: str) -> ResultDict:
    """
    Inspects migration status (applied vs pending) against any standard DB-API 2.0 connection.
    """
    try:
        cursor = conn.cursor()
        init_sql = """
        CREATE TABLE IF NOT EXISTS schema_migrations (
            version VARCHAR(255) PRIMARY KEY,
            applied_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
        );
        """
        cursor.execute(init_sql)
        conn.commit()

        cursor.execute("SELECT version FROM schema_migrations;")
        applied_set: Set[str] = {str(row[0]) for row in cursor.fetchall()}

        if not os.path.exists(migrations_dir):
            raise FileNotFoundError(f"Migrations directory not found: {migrations_dir}")

        entries = []
        for filename in os.listdir(migrations_dir):
            if not filename.endswith(".sql"):
                continue
            match = VERSION_REGEX.match(filename)
            if match:
                version = match.group(1)
                name = match.group(2) or filename[:-4]
                entries.append(ResultDict({
                    "version": version,
                    "name": name,
                    "applied": version in applied_set,
                    "applied_at": None,
                }))

        entries.sort(key=lambda x: x["version"])

        return ResultDict({
            "success": True,
            "migrations": entries,
            "error": None,
        })
    except Exception as e:
        return ResultDict({
            "success": False,
            "migrations": [],
            "error": str(e),
        })


def detect_dialect(conn: Any) -> str:
    """
    Detects the SQL dialect ('postgres', 'mysql', or 'sqlite') based on the connection type.
    Defaults to 'postgres' if unknown.
    """
    if conn is None:
        return "postgres"
    conn_type = f"{type(conn).__module__}.{type(conn).__name__}".lower()
    if "sqlite" in conn_type:
        return "sqlite"
    if "mysql" in conn_type:
        return "mysql"
    return "postgres"


def _get_engine_funcs():
    try:
        from .migration_engine_py import diff_drawdb, plan_sync_drawdb
        return diff_drawdb, plan_sync_drawdb
    except (ImportError, ValueError):
        try:
            from migradb.migration_engine_py import diff_drawdb, plan_sync_drawdb
            return diff_drawdb, plan_sync_drawdb
        except ImportError:
            from migration_engine_py import diff_drawdb, plan_sync_drawdb
            return diff_drawdb, plan_sync_drawdb


def sync_database(
    conn: Any,
    schema: str = "./schema/drawdb.json",
    migrations_dir: str = "./migrations",
    schema_dir: str = "./schema",
    dialect: Optional[str] = None,
    save_migration_file: bool = True,
    migration_name: str = "sync_drawdb",
    force_full: bool = False,
) -> ResultDict:
    """
    Synchronizes the database schema against a DrawDB schema file.
    Dynamically compares schema against snapshot, translates DDL for target dialect,
    executes in a transaction, updates schema_migrations and snapshot, and optionally saves audit SQL.
    """
    import datetime

    target_dialect = dialect or detect_dialect(conn)

    os.makedirs(schema_dir, exist_ok=True)
    os.makedirs(migrations_dir, exist_ok=True)

    diff_drawdb, plan_sync_drawdb = _get_engine_funcs()

    plan = plan_sync_drawdb(
        schema_path=schema,
        migrations_dir=migrations_dir,
        schema_dir=schema_dir,
        dialect=target_dialect,
        force_full=force_full,
    )

    if not plan.success and plan.error:
        return ResultDict({
            "success": False,
            "is_empty": False,
            "applied": 0,
            "diff_summary": "",
            "migration_path": None,
            "statements": [],
            "error": plan.error,
        })

    if plan.is_empty or not plan.statements:
        return ResultDict({
            "success": True,
            "is_empty": True,
            "applied": 0,
            "diff_summary": plan.diff_summary,
            "statements": [],
            "migration_path": None,
            "error": None,
        })

    try:
        cursor = conn.cursor()

        init_sql = """
        CREATE TABLE IF NOT EXISTS schema_migrations (
            version VARCHAR(255) PRIMARY KEY,
            applied_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
        );
        """
        cursor.execute(init_sql)
        conn.commit()

        # Execute statements
        for stmt in plan.statements:
            s = stmt.strip()
            if s:
                cursor.execute(s)

        now = datetime.datetime.utcnow().strftime("%Y%m%d%H%M%S")
        record_sql = f"INSERT INTO schema_migrations (version) VALUES ('{now}');"
        cursor.execute(record_sql)
        conn.commit()

        migration_path = None
        if save_migration_file:
            diff_res = diff_drawdb(
                schema_path=schema,
                migrations_dir=migrations_dir,
                schema_dir=schema_dir,
                migration_name=migration_name,
                dialect=target_dialect,
                force_full=force_full,
            )
            if diff_res and diff_res.migration_path:
                migration_path = diff_res.migration_path
        else:
            with open(schema, "r", encoding="utf-8") as f:
                content = f.read()
            with open(os.path.join(schema_dir, "drawdb_snapshot.json"), "w", encoding="utf-8") as f:
                f.write(content)
            with open(os.path.join(migrations_dir, ".schema_snapshot.json"), "w", encoding="utf-8") as f:
                f.write(content)

        return ResultDict({
            "success": True,
            "is_empty": False,
            "applied": len(plan.statements),
            "diff_summary": plan.diff_summary,
            "migration_path": migration_path,
            "statements": list(plan.statements),
            "error": None,
        })
    except Exception as e:
        try:
            conn.rollback()
        except Exception:
            pass
        return ResultDict({
            "success": False,
            "is_empty": False,
            "applied": 0,
            "diff_summary": plan.diff_summary,
            "migration_path": None,
            "statements": list(plan.statements),
            "error": str(e),
        })

File: npm/runner.js
const fs = require('fs');
const path = require('path');

const VERSION_REGEX = /^(\d+)(?:_(.+))?\.sql$/;

/**
 * Executes an SQL query against different types of database clients (pg, mysql2, sqlite, or callback).
 */
async function executeQuery(client, sqlQuery) {
  if (typeof client === 'function') {
    return await client(sqlQuery);
  }
  // pg / mysql2 style
  if (typeof client.query === 'function') {
    return await client.query(sqlQuery);
  }
  // better-sqlite3 style
  if (typeof client.exec === 'function') {
    return client.exec(sqlQuery);
  }
  // sqlite3 callback style
  if (typeof client.run === 'function') {
    return new Promise((resolve, reject) => {
      client.run(sqlQuery, (err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }
  throw new Error('Unsupported database client. Expected client with query(), exec(), or run(), or a query function.');
}

/**
 * Fetches applied versions from schema_migrations table.
 */
async function getAppliedVersions(client) {
  const query = 'SELECT version FROM schema_migrations;';
  if (typeof client === 'function') {
    const res = await client(query);
    return extractVersions(res);
  }
  if (typeof client.query === 'function') {
    const res = await client.query(query);
    // pg returns res.rows, mysql returns [rows, fields]
    const rows = res && res.rows ? res.rows : (Array.isArray(res) ? res[0] : res);
    return extractVersions(rows);
  }
  if (typeof client.prepare === 'function') {
    // better-sqlite3
    const rows = client.prepare(query).all();
    return extractVersions(rows);
  }
  return new Set();
}

function extractVersions(rows) {
  const set = new Set();
  if (Array.isArray(rows)) {
    for (const r of rows) {
      if (r && r.version) {
        set.add(String(r.version));
      }
    }
  }
  return set;
}

/**
 * Runs all pending migrations against the provided database connection.
 * @param {object|Function} client - Database connection (pg, mysql2, better-sqlite3, or query callback)
 * @param {string} migrationsDir - Path to directory containing .sql migration files
 */
async function runOnDatabase(client, migrationsDir) {
  try {
    // 1. Ensure schema_migrations table exists
    const initSql = `
      CREATE TABLE IF NOT EXISTS schema_migrations (
        version VARCHAR(255) PRIMARY KEY,
        applied_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
      );
    `;
    await executeQuery(client, initSql);

    // 2. Fetch applied migrations
    const appliedSet = await getAppliedVersions(client);

    // 3. Scan migrations directory
    if (!fs.existsSync(migrationsDir)) {
      throw new Error(`Migrations directory not found: ${migrationsDir}`);
    }

    const files = fs.readdirSync(migrationsDir);
    const entries = [];

    for (const f of files) {
      if (!f.endsWith('.sql')) continue;
      const match = f.match(VERSION_REGEX);
      if (match) {
        entries.push({
          version: match[1],
          name: match[2] || f.replace(/\.sql$/, ''),
          filename: f,
        });
      }
    }

    entries.sort((a, b) => (a.version > b.version ? 1 : -1));

    const applied = [];
    for (const entry of entries) {
      if (appliedSet.has(entry.version)) {
        continue;
      }

      const filePath = path.join(migrationsDir, entry.filename);
      const sqlContent = fs.readFileSync(filePath, 'utf8');

      if (sqlContent.trim().length > 0) {
        await executeQuery(client, sqlContent);
      }

      // Record migration
      const recordSql = `INSERT INTO schema_migrations (version) VALUES ('${entry.version}');`;
      await executeQuery(client, recordSql);

      applied.push({
        version: entry.version,
        name: entry.name,
      });
    }

    return {
      success: true,
      applied: applied.length,
      migrations: applied,
    };
  } catch (err) {
    return {
      success: false,
      applied: 0,
      migrations: [],
      error: err.message || String(err),
    };
  }
}

/**
 * Inspects migration status against the database without applying changes.
 */
async function statusOnDatabase(client, migrationsDir) {
  try {
    const initSql = `
      CREATE TABLE IF NOT EXISTS schema_migrations (
        version VARCHAR(255) PRIMARY KEY,
        applied_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
      );
    `;
    await executeQuery(client, initSql);
    const appliedSet = await getAppliedVersions(client);

    if (!fs.existsSync(migrationsDir)) {
      throw new Error(`Migrations directory not found: ${migrationsDir}`);
    }

    const files = fs.readdirSync(migrationsDir);
    const list = [];

    for (const f of files) {
      if (!f.endsWith('.sql')) continue;
      const match = f.match(VERSION_REGEX);
      if (match) {
        list.push({
          version: match[1],
          name: match[2] || f.replace(/\.sql$/, ''),
          applied: appliedSet.has(match[1]),
        });
      }
    }

    list.sort((a, b) => (a.version > b.version ? 1 : -1));

    return {
      success: true,
      migrations: list,
    };
  } catch (err) {
    return {
      success: false,
      migrations: [],
      error: err.message || String(err),
    };
  }
}

function detectDialect(client) {
  if (!client) return 'postgres';
  if (typeof client.prepare === 'function' || (typeof client.exec === 'function' && !client.query)) {
    return 'sqlite';
  }
  if (client.threadId !== undefined || client.config?.database || (client.format && client.escape)) {
    return 'mysql';
  }
  return 'postgres';
}

/**
 * Synchronizes the database schema against a DrawDB schema file.
 * Dynamically compares schema against snapshot, translates DDL for target dialect,
 * executes in transaction, updates schema_migrations and snapshot, and optionally saves audit SQL.
 */
async function syncDatabase(client, options = {}) {
  const {
    schema = './schema/drawdb.json',
    migrationsDir = './migrations',
    schemaDir = './schema',
    dialect: specifiedDialect,
    saveMigrationFile = true,
    migrationName = 'sync_drawdb',
    forceFull = false,
  } = options;

  const dialect = specifiedDialect || detectDialect(client);

  if (!fs.existsSync(schemaDir)) {
    fs.mkdirSync(schemaDir, { recursive: true });
  }
  if (!fs.existsSync(migrationsDir)) {
    fs.mkdirSync(migrationsDir, { recursive: true });
  }

  const { planSyncDrawdb, diffDrawdb } = require('./index');

  const plan = planSyncDrawdb(schema, migrationsDir, schemaDir, dialect, forceFull);
  if (!plan.success && plan.error) {
    return {
      success: false,
      isEmpty: false,
      applied: 0,
      diffSummary: '',
      migrationPath: null,
      statements: [],
      error: plan.error,
    };
  }

  if (plan.isEmpty || !plan.statements || plan.statements.length === 0) {
    return {
      success: true,
      isEmpty: true,
      applied: 0,
      diffSummary: plan.diffSummary,
      statements: [],
      migrationPath: null,
    };
  }

  const initSql = `
    CREATE TABLE IF NOT EXISTS schema_migrations (
      version VARCHAR(255) PRIMARY KEY,
      applied_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
    );
  `;
  await executeQuery(client, initSql);

  try {
    await executeQuery(client, 'BEGIN;');
    for (const stmt of plan.statements) {
      const trimmed = stmt.trim();
      if (trimmed) {
        await executeQuery(client, trimmed);
      }
    }

    const now = new Date();
    const pad = (n) => String(n).padStart(2, '0');
    const version = `${now.getUTCFullYear()}${pad(now.getUTCMonth() + 1)}${pad(now.getUTCDate())}${pad(now.getUTCHours())}${pad(now.getUTCMinutes())}${pad(now.getUTCSeconds())}`;
    const recordSql = `INSERT INTO schema_migrations (version) VALUES ('${version}');`;
    await executeQuery(client, recordSql);

    await executeQuery(client, 'COMMIT;');

    let migrationPath = null;
    if (saveMigrationFile) {
      const diffRes = diffDrawdb(schema, migrationsDir, schemaDir, migrationName, dialect, forceFull);
      if (diffRes && diffRes.migrationPath) {
        migrationPath = diffRes.migrationPath;
      }
    } else {
      const content = fs.readFileSync(schema, 'utf8');
      fs.writeFileSync(path.join(schemaDir, 'drawdb_snapshot.json'), content);
      fs.writeFileSync(path.join(migrationsDir, '.schema_snapshot.json'), content);
    }

    return {
      success: true,
      isEmpty: false,
      applied: plan.statements.length,
      diffSummary: plan.diffSummary,
      migrationPath,
      statements: plan.statements,
    };
  } catch (err) {
    try {
      await executeQuery(client, 'ROLLBACK;');
    } catch (_) {}
    return {
      success: false,
      isEmpty: false,
      applied: 0,
      diffSummary: plan.diffSummary,
      migrationPath: null,
      statements: plan.statements,
      error: err.message || String(err),
    };
  }
}

module.exports = {
  runOnDatabase,
  statusOnDatabase,
  syncDatabase,
  detectDialect,
};

